### PR TITLE
Fix interesting outlier where the only available land spaces are hazards. And you're broke.

### LINKED ISSUE #1823: [Ares] Bug - placing tile with no placement possibilities
Hi,
during my game I played Research Outpost card. Only places without tiles next to it were Dust Storms, which require 8 / 16 $ to place there. However I do not have that much money. Currently I'm stuck inside this action.
Here is this my turn: https://terraforming-mars.herokuapp.com/player?id=a862fa2596d2

Here are some screenshots (im not sure if link above will work):
Screenshot from 2020-11-04 00-25-24
Screenshot from 2020-11-04 00-25-34

Thanks for your awesome work!

### PR DESCRIPTION
**Pull request by kberg**
Is this enough to c.l.o.s.e #1823?

### CODE REVIEW

File: src/Board.ts [hunk 1]
```diff
@@ -143,16 +143,53 @@ export abstract class Board {
...
+                return true;
+            }
+
+            // b) the space has a regular tile.
+            if (!AresHandler.hasHazardTile(space)) {
+                return false;
+            }
+
+            // c) the space has a hazard tile.
+            if (AresHandler.hasHazardTile(space)) {
```
Comment (Lynesth) [inline src/Board.ts:168]: Can it be anything else at this point ?

File: src/Board.ts [hunk 2]
```diff
@@ -143,16 +143,53 @@ export abstract class Board {
...
+                // Hazard tiles can be covered if they're not protected ... 
+                const blockedByDesperateMeasures = space.tile?.protectedHazard === true;
```
Comment (Lynesth) [inline src/Board.ts:170]: You can probably leave the `?` out since `space.tile` cannot be undefined at this point.

File: src/Board.ts [hunk 3]
```diff
@@ -143,16 +143,53 @@ export abstract class Board {
             );
     }
 
-    public getAvailableSpacesOnLand(player?: Player): Array<ISpace> {
+    public getAvailableSpacesOnLand(player?: Player, checkAffordability = true): Array<ISpace> {
```
Comment (Lynesth) [inline src/Board.ts:146]: Set `checkAffordability` type to boolean.
Also, when is `getAvailableSpacesOnLand` ever called with `false` as a second argument ? Does this need to be an argument ?

File: src/Board.ts [hunk 4]
```diff
@@ -143,16 +143,53 @@ export abstract class Board {
...
+            if (!AresHandler.hasHazardTile(space)) {
+                return false;
+            }
+
+            // c) the space has a hazard tile.
+            if (AresHandler.hasHazardTile(space)) {
+                // Hazard tiles can be covered if they're not protected ... 
+                const blockedByDesperateMeasures = space.tile?.protectedHazard === true;
+                // and the player can pay the coverage cost
+                const canAfford = (checkAffordability &&
```
Comment (Lynesth) [inline src/Board.ts:172]: So if `checkAffordability` is set to false, we can't afford it ?

**Comment by Lynesth** [comment]
Should have grouped those comments into a review, sorry about that.

File: src/Board.ts (hunk 1 above)
Comment (kberg) [reply src/Board.ts:168]: Could be an ocean on a land space.

File: src/Board.ts (hunk 2 above)
Comment (kberg) [reply src/Board.ts:170]: Oh good! Fixed.

File: src/Board.ts (hunk 4 above)
Comment (kberg) [reply src/Board.ts:172]: Whoops! Changed that. I'll add a test. :D Not that I honestly think anyone will call that will checkAffordability === false.

Comment (kberg) [reply src/Board.ts:172]: Perhaps if it's the neutral player?

File: src/Board.ts (hunk 1 above)
Comment (Lynesth) [reply src/Board.ts:168]: That would be caught by the previous if block
```ts
             // b) the space has a regular tile. 
             if (!AresHandler.hasHazardTile(space)) { 
```

File: src/Board.ts [hunk 5]
```diff
@@ -167,9 +167,9 @@ export abstract class Board {
             // c) the space has a hazard tile.
             if (AresHandler.hasHazardTile(space)) {
                 // Hazard tiles can be covered if they're not protected ... 
-                const blockedByDesperateMeasures = space.tile?.protectedHazard === true;
+                const blockedByDesperateMeasures = space.tile.protectedHazard === true;
                 // and the player can pay the coverage cost
-                const canAfford = (checkAffordability && 
+                const canAfford = (!checkAffordability ||
                     (player === undefined || player.canAfford(AresHandler.hazardCost(space))));
```
Comment (Lynesth) [inline src/Board.ts:173]: You can remove extra parenthesis not needed anymore here.

**Comment by Lynesth** [comment]
Actually, this PR does not take into account what the player must pay for the card/action as well and will return false positives and errors.

Example using Research Outpost (costs 18 MC). Let's say the player has 20 MC and all available spaces for the city are hazards.
Here's the card's canPlay method:
```ts
    public canPlay(player: Player, game: Game): boolean {
        return this.getAvailableSpaces(player, game).length > 0;
    }
```
This will return true but once the card is played (and the player now only has 2 MC), it will throw an error has it will pass an empty spaces list to the SelectSpace input.

**Comment by kberg** [comment]
> Actually, this PR does not take into account what the player must pay for the card/action as well and will return false positives and errors.
> 
> Example using Research Outpost (costs 18 MC). Let's say the player has 20 MC and all available spaces for the city are hazards.
> Here's the card's canPlay method:
> 
> ```ts
>     public canPlay(player: Player, game: Game): boolean {
>         return this.getAvailableSpaces(player, game).length > 0;
>     }
> ```
> 
> This will return true but once the card is played (and the player now only has 2 MC), it will throw an error has it will pass an empty spaces list to the SelectSpace input.

That's a good observation. Damn.

You said you had a larger fix?

**Pull request closed without merging**